// ufo.rpi/storageInfo.js
var njds = require('nodejs-disks');

var USB_MOUNTPT = '/media/usb';


var KB = 1024;
var MB = 1024 * KB;
var GB = 1024 * MB;
var TB = 1024 * GB;

function bytes(str) {
    var base = parseFloat(str);
    var units = [['KB', KB], ['MB', MB], ['GB', GB], ['TB', TB]];
    for (var i = 0; i < units.length; i++) {
        var unit = units[i];
        if (str.indexOf(unit[0]) > 0) {
            return Math.round(base * unit[1]);
        }
    }
    return base;
}

function storageInfo(cb) {
    njds.drives(function (err, drives) {
        if (err != null) {
            cb(err);
        }
        else {
            njds.drivesDetail(
                drives,
                function (err, data) {
                    if (err != null) {
                        cb(err);
                    }
                    else {
                        var usbInf = data.filter(function(v){return v.mountpoint == USB_MOUNTPT});
                        if (usbInf.length == 0) {
                            cb(null, {
                                mounted: false
                            });
                        }
                        else {
                            usbInf = usbInf[0];
                            cb(null, {
                                mounted: true,
                                total: bytes(usbInf.total),
                                used: bytes(usbInf.used),
                                available: bytes(usbInf.available)
                            });
                        }
                    }
                })
            }
        });
}

module.exports = {
    MOUNTPT: USB_MOUNTPT,
    storageInfo: storageInfo
};
